### ISSUE TITLE
Fix StatefulSet volume provisioning "magic"

### ISSUE THREAD

**Issue by kow3ns** [kubernetes/kubernetes#41598]
**Feature Request** 

**tldr** 
Currently the volume [utilities](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util.go#L287) used by the GCE and AWS cloud providers attempt to spread the Pods of StatefulSest across zones by detecting PVCs whose names conform to the StatefulSet naming scheme and round robin provisioning the required PVs across AWS and GCE zones. Aside from introducing “magic” for StatefulSet, which makes it not forkable, it does not currently work as documented, and its interaction with Pod Anti-Affinity will probably be, at best, surprising to users.

**Details**
StatefulSet creates PVCs prior to creating its Pods. For many Cloud Providers, the PVs, which are bound to the StatefulSets PVCs, are zonal resources. In order to achieve a high availability deployment, and to ensure the durability of data in the event of a zone failure, the PVs that correspond to PVCs requested by the StatefulSet controller are provisioned across zones in a round robin fashion by Volume Plugins for those environments.  After this, the Pods of the StatefulSet are bound to the zones where the PVs have been provisioned.
1. Zonal anti-affinity will not work as described for the Pods of a StatefulSet on cloud providers that have zonal PVs (i.e AWS, GCP, GKE). Once the PVCs are bound to a PV in a particular zone, the StatefulSet Pod must be scheduled in that zone. Specifying a hard, zonal anti-affinity can still be used to ensure that users will never have more than one Pod per zone, but soft zonal anti-affinity can not be used to achieve spreading across zones. 
2. The implementation will not correctly collocate PVCs from the same StatefulSet Pod in the same AZ/Zone. This causes the victim Pod to be unschedulable #35695.
3. The current implementation applies round robin spreading to any PVCs that end with -id, where id is a number. This will not exclude PVCs that were not generated by a StatefulSet. Given the naming constraints of PVC, it is impossible that users will use names like my-app-pvc-1, my-app-pvc-2, etc.
4. Attempts to fix 2. by grouping by the Pod name and spreading on the ordinal can not be successful in the general case #40910. PVCs created by StatefulSet Pods can not be identified by name. The current construction is ambiguous.
```raw
. Ordinal -> [0-9]+
. Subdomain ->  [a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*
. SetName -> Subdomain
. ClaimName -> Subdomain
. StatefulSetClaimName -> ClaimName-SetName-Ordinal
```
The legal name foo-bar-foo-bar-1 admits a variety of valid parses. Also, even if we take this approach, we are really only adding more StatefulSet "magic"

At a minimum, we need to document the current behavior. In general, as the relationship between storage affinity/anti-affinity and cluster topology (e.g. host, zone, rack, power domain) will become even complicated and critical with [local PVs](https://github.com/kubernetes/community/pull/306), a more robust solution is probably necessary.

**Comment by kow3ns** [comment]
@kubernetes/sig-storage-feature-requests 
@kubernetes/sig-scheduling-feature-requests

**Comment by smarterclayton** [comment]
We talked in another proposal about attaching affinity selectors to PVCs and letting the tail wag the dog.  That might work for a StatefulSet, but probably not useful for the other controllers.

**Comment by davidopp** [comment]
ref/ #41442

**Comment by justinsb** [comment]
I agree we should replace the heuristics with the scheduler.  That is why the heuristics are quite so fragile - we didn't want to paint the scheduler into a corner.  We discussed adding additional metadata to the PVC, but the fear was that it would interact badly with the long term plan.

However, I think we're technically going to resolve this issue when I clean up #40910:

1. remains the same whether the heuristics are in place or not - once the PV is created in a zone, you have no choice as to the pod's zone placement.
2. will be fixed in #40910 
3. Is a feature to influence zone placement for dynamic volumes for users doing something statefulset like without using statefulsets.  (It's better than random)  I agree documentation would be nice, but if we document it we make life harder for the scheduler.
4. will be fixed in #40910

I'm not sure how to treat this issue therefore.  Presumably you don't want it closed if/when #40910 merges @kow3ns , but technically as it stands today it probably should be.

I think perhaps opening a feature in the features repo with your design of the desired behaviour is a good next step?  I believe the status-quo end-goal is that the scheduler schedules volumes to zones, considering the requirements of the associated pods and available resources.  As well as pure implementation challenges, there is also the challenge that currently the scheduling information (e.g. tolerations) are on the pod, not the PVC and the link is not necessarily known when we create the PV, hence @smarterclayton's suggestion (I think).

**Comment by kow3ns** [comment]
@justinsb 
1. At a minimum we need to be clear with users about current status of dynamic volume provisioning on AWS or GCP (even with #40910). The tldr is that we really don't have a way to achieve zonal collocation or controlled spreading for stateful applications that use dynamically provisioned storage on these providers, at this time. I think we would have to document the following to clearly express this limitation. 
- If you create a PVC with an name that matches (.*)-([0-9]+)$, the provisioner will round-robin provision volumes for those claims across zones in the region.
- If the claims created do not strictly match the naming scheme implemented by StatefulSet, or if they are created by a StatefulSet whose name contains dashes, provisioning may be non-uniform across zones. Under these circumstances, users should not depend on the provisioner to provide a uniform distribution of storage across a region's zones.
- If users mount the volumes created by the dynamic provisioner, and if, via a StatefulSet's PodTemplateSpec or some other means, a hard or zonal affinity or a hard zonal anti-affinity is applied to the Pods to which these volumes are mounted, it may cause those Pods to become unschedulable.  Therefore, when using the dynamic provisioner on AWS or GCP, users should not attempt to achieve zonal co-location or spreading using these fields.
- Under the same conditions as above, if you use soft zonal affinity, or soft zonal anti-affinity it will have no effect. Users should not depend on these fields to achieve zonal collocation or spreading.

2. I think that a next step, prior to a design, is for apps, storage, and scheduling to get together and agree on the outcome we want to deliver. Right now the proivisioner is effectively scheduling Pods at a zonal level, and maybe that's okay. The provisioner might be the right place to encapsulate the concern of topology constraints for remote storage (though if it is we need some more knobs). However, given that PVCs and PVs are going to be used for local storage, we might need to make the scheduler aware of storage constraints anyway (otherwise you can easily provision a bunch of local storage on a host that doesn't meet your mem, cpu, etc requirements).

**Comment by justinsb** [comment]
Totally agree on both points.  For the docs I think in general we should describe it as best-effort spreading rather than a guarantee anyway, because we _also_ use heuristics to find the set of zones.  (And yes, we had a edge-case here also  - #41702 ).  If you want to take your bullet points and put them into a google doc or PR for the docs I can review/comment and/or iterate on them if that would be helpful.

The underlying problem is non-trivial though.  For example, if an AZ is temporarily unavailable, where should we create volumes for a ScheduledSet?  #38505 is one approach, but there's also a difference between a 2 node active/passive statefulset in a 3 zone environment (skip the failed zone?) vs a 3 node quorum-style statefulset in a 3 zone environment (bring up nodes in the active zones, but don't put 2 nodes into the same zone - wait for the 3rd zone to recover?)  I suspect the answer is to make affinity work, but I don't know.

I definitely agree on the groups working together to figure this all out, and am happy to help.  It isn't entirely clear to me what the end-goal looks like, other than that it won't involve these heuristics ;-)  But I still believe we do need these heuristics (or equivalent behaviour) in the meantime - I'm just sorry I messed up the multi-volume piece.

**Comment by bbeaudreault** [comment]
I'd like to echo @kow3ns recent comment. The existing heuristic fails when the StatefulSet name has dashes in it. This is rather fragile considering dashes are the *only allowed separators* in kubernetes names. I've been bitten by this recently and had to remove all dashes from the name, which results in harder to reason about organization. The specific pod in question is sharded mysql, the statefulset name is `dbname-shard-version`. This results in a pvc name of `logs-dbname-0-0-0`, so the spreading is happening on the version number (second to last 0) instead of the statefulset name `dbname-0-0`. Removing dashes results in `dbname00` which is harder to read but "fine" at low numbers. It becomes much more ambiguous when you imagine `dbname110` which could be shard 1 version 10 or shard 11 version 0

**Comment by smarterclayton** [comment]
Agree, putting this on the 1.7 radar for now.

**Comment by justinsb** [comment]
@bbeaudreault what are the symptoms that you are seeing?  Suboptimal spreading, or an actual failure?  If you stateful set name is `dname-shared-version` then we will seed the spreading based on `version`, but should still spread across zones.

**Comment by bcorijn** [comment]
As I see little remark about other controllers: from my experience this issue does not just hit StatefulSets, a Deployment which mounts 2 PVC's to one pod can also get blocked because the PV's get spread out across zones.

**Comment by justinsb** [comment]
@bcorijn yes.  For now you can use a storage class, or you can abuse the heuristics to work in your favor.  But we're all eagerly waiting for the real solution :-)

**Comment by kkurzhal** [comment]
I wasn't sure if this was mentioned anywhere else, or if anyone else had thought about/discussed this, but this issue seemed to be the best place to put it.

There is an implied problem regarding resource requests due to the way that statefulsets are scheduled.  If the pvc is "blindly allocated" to a zone without first checking the node resources that are available to the statefulset then it is possible that a zone will be selected that does not have enough resources, which will prevent the pod from being scheduled even if there are available resources in other zones.  And since the statefulset pod must always be scheduled in the same zone as the matching pvc, this means that the scheduling is halted.

My team and I ran into this issue this week when trying to schedule an elasticsearch cluster based on https://github.com/pires/kubernetes-elasticsearch-cluster.  We were able to circumvent this by making sure that the statefulset was created before the clients; but as a user, it was a bit of an unexpected issue and a little hard to determine why I was receiving "Insufficient cpu, NoVolumeZoneConflict" errors.

Please let me know if you'd like for me to create a new issue for this, or if you'd like for me to post my configs/cluster info that we had an issue with.

**Comment by smarterclayton** [comment]
That problem is definitely important - it sits at the overlap between
sig-scheduling (who need to be able to enable the correct choices) and both
sig-apps and sig-storage.

**Comment by msau42** [comment]
I had an idea for how to integrate storage topology into the scheduler.  I outlined it in #43504

**Comment by justinsb** [comment]
Something that has occurred to me ... autoscaling turns this on its head.  On the cloud, we should probably optimize for spreading, and then launch additional instances as needed.  i.e. if we're configured to run in zones A,B and C but happen only to have instances in A & B, we should still create a volume in zone C, and then spin up an instance in zone C.

**Comment by k82cn** [comment]
move to 1.8?

**Comment by kow3ns** [comment]
@k82cn yes please

**Comment by msau42** [comment]
@kow3ns I am working on integrating PV binding into pod scheduling, so that the PV binding decision can be influenced by pod anti-affinity, resource requirements, node resource availability, etc.  So once I can get that working with dynamic provisioning and zonal volumes (GCE PD, AWS EBS, etc), then the StatefulSet spreading can be driven from the pod constraints, rather than from the volume provisioning.  It's still going to be a few more releases away before it's fully functional though.

**Comment by davidopp** [comment]
+1 for trying to solve this together with the local PV scheduling.

**Comment by krmayankk** [comment]
@msau42 i am happy to help once this reaches that stage

**Comment by airstand** [comment]
+1

**Comment by VinceMD** [comment]
+1 I think this is somewhat mentioned here for Cassandra example https://github.com/kubernetes/examples/issues/149 .

Scheduling can be halted for as long as the first zone used in a statefulset is down. We should let Kubernetes use the other zones that are still healthy.

**Comment by msau42** [comment]
/remove-lifecycle stale

Working on the design for dynamic provisioning now.

**Comment by msau42** [comment]
Dynamic provisioning topology design proposal is here: kubernetes/community#1857

**Comment by shilpamayanna** [comment]
+1

**Comment by AndresPineros** [comment]
Still having this bug at 1.9

**Comment by marandalucas** [comment]
We have the same problem when we deploy kafka "statefulset" with dynamic pvc.
_The pod is always in Pending..._

kube-scheduler logs:
`I0808 08:57:17.649951       1 scheduler.go:191] Failed to schedule pod: namespace/namespace-kafka-svc-kafka-0
I0808 08:57:17.650022       1 factory.go:1251] Updating pod condition for namespace/namespace-kafka-svc-kafka-0 to (PodScheduled==False)
I0808 08:57:33.652507       1 scheduler.go:191] Failed to schedule pod: namespace/namespace-kafka-svc-kafka-0
I0808 08:57:33.652631       1 factory.go:1251] Updating pod condition for namespace/namespace-kafka-svc-kafka-0 to (PodScheduled==False)`
In this moment we have the followings nodes (The pv's are in eu-west-1a ):
- 2 nodes eu-west-1a
- 3 nodes eu-west-1b
- 3 nodes eu-west-1c

We've been reading some issues about it for two days and we have solved it by putting the same number of nodes in each of the availability zones.
- 3 nodes eu-west-1a
- 3 nodes eu-west-1b
- 3 nodes eu-west-1c

**Comment by msau42** [comment]
Topology-aware dynamic provisioning for the gce-pd, aws-ebs and azure disks is available as beta in 1.12

**Comment by msau42** [comment]
Is there additional work that should be done for StatefulSets when using topology-aware dynamic provisioning to retain the default "zone spreading" behavior?  Or do we require users to explicitly set zone ant-affinity in their StatefulSets now?
